lib/ecto/query.ex

```diff
@@ -1071,6 +1071,11 @@ defmodule Ecto.Query do
                            "list of variables, got: #{Macro.to_string(binding)}"
   end
 
+  defmacro join(_query, _qual, binding, _expr, opts) when is_list(binding) do
```

Comment:
Cool! But what about if both `binding` and `opts` are not lists? Shouldn't we throw better exceptions. As far as I understand here, in that case we would throw a `FunctionClauseError` 🤔

Reply:
There's already a check for that.

Reply:
Are you sure? I couldn't find it, the other clauses match for when both `binding` and `opts` are lists or when only `opts` is a list. You added a clause for when only `binding` is a list. If both are not lists, no clause will match and one will get a `FunctionClauseError`

Reply:
Ah gotcha! Latest commit should cover it (dev would get one error to fix, and then the other).